airbyte-cdk/bulk/core/extract/src/main/kotlin/io/airbyte/cdk/read/StateManager.kt

```diff
@@ -85,9 +85,14 @@ class StateManager(
      * Returns the Airbyte STATE messages which checkpoint the progress of the READ in the platform.
      * Updates the internal state of the [StateManager] to ensure idempotency (no redundant messages
      * are emitted).
+     *
+     * Note to avoid multithreading causing record count to be off, we'd only report streams with
+     * records in it. Any streams without stream will be delayed to report until next checkpoint.
+     * This is based on assumption each stream will be mapped to one nonGlobal manager.
      */
     fun checkpoint(): List<AirbyteStateMessage> =
-        listOfNotNull(global?.checkpoint()) + nonGlobal.mapNotNull { it.value.checkpoint() }
+        listOfNotNull(global?.checkpoint()) +
+            nonGlobal.mapNotNull { it.value.checkpoint() }.filter { it.sourceStats.recordCount > 0 }
```

Comment:
We should instead change the implementation of `checkpoint()` for `NonGlobalStreamStateManager` so that it thinks the state value is `Stale` in the case you're trying to cover.

Reply:
Scratched this as I didn't know we already have 'Stale' mechanism. However I found a bug there and that might contribute to the problem. Please see my latest diff!